node/zombienet-backchannel/src/lib.rs

```diff
@@ -0,0 +1,155 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Provides the possibility to coordination between malicious actors and
+//! the zombienet test-runner, allowing to reference runtime's generated
+//! values in the test specifications, through a bidirectional message passing
+//! implemented as a `backchannel`.
+
+use futures_util::{SinkExt, StreamExt};
+use lazy_static::lazy_static;
+use parity_scale_codec as codec;
+use serde::{Deserialize, Serialize};
+use std::{env, sync::Mutex};
+use tokio::sync::broadcast;
+use tokio_tungstenite::{connect_async, tungstenite::protocol::Message};
+
+mod errors;
+use errors::BackchannelError;
+
+lazy_static! {
+	pub static ref ZOMBIENET_BACKCHANNEL: Mutex<Option<ZombienetBackchannel>> = Mutex::new(None);
+}
+
+#[derive(Debug)]
+pub struct ZombienetBackchannel {
+	broadcast_tx: broadcast::Sender<BackchannelItem>,
+	ws_tx: broadcast::Sender<BackchannelItem>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct BackchannelItem {
+	key: String,
+	value: String,
+}
+
+pub struct Broadcaster;
+
+pub const ZOMBIENET: &str = "🧟ZOMBIENET🧟";
+
+impl Broadcaster {
+	/// Return a subscriber that will receive all message broadcasted by the zombienet backchannel
+	/// websocket server.
+	pub fn subscribe(&self) -> Result<broadcast::Receiver<BackchannelItem>, BackchannelError> {
+		let mut zombienet_bkc = ZOMBIENET_BACKCHANNEL.lock().unwrap();
+		let sender = zombienet_bkc.as_mut().unwrap().broadcast_tx.clone();
+		if zombienet_bkc.is_some() {
+			Ok(sender.subscribe())
+		} else {
+			Err(BackchannelError::Uninitialized)
+		}
+	}
+
+	/// Provides a simple API to send a key/value to the zombienet websocket server.
+	pub async fn send(
+		&mut self,
+		key: &'static str,
+		val: impl codec::Encode,
+	) -> Result<(), BackchannelError> {
+		let mut zombienet_bkc = ZOMBIENET_BACKCHANNEL.lock().unwrap();
+		if zombienet_bkc.is_none() {
+			return Err(BackchannelError::Uninitialized)
+		}
+		let encoded = val.encode();
+		let backchannel_item = BackchannelItem {
+			key: key.to_string(),
+			value: String::from_utf8_lossy(&encoded).to_string(),
+		};
+
+		let sender = zombienet_bkc.as_mut().unwrap().ws_tx.clone();
+		sender.send(backchannel_item).map_err(|e| {
+			tracing::error!(target = ZOMBIENET, "Error sending new item: {}", e);
+			BackchannelError::SendItemFail
+		})?;
+
+		Ok(())
+	}
+}
+
+impl ZombienetBackchannel {
+	pub async fn init() -> Result<(), BackchannelError> {
+		let mut zombienet_bkc = ZOMBIENET_BACKCHANNEL.lock().unwrap();
+		if zombienet_bkc.is_none() {
+			let backchannel_host =
+				env::var("BACKCHANNEL_HOST").unwrap_or_else(|_| "backchannel".to_string());
+			let backchannel_port =
+				env::var("BACKCHANNEL_PORT").unwrap_or_else(|_| "3000".to_string());
```

Comment:
I'd recommend parsing the content of those first, before assuming their content is sane, and only replace the parsed values by defaults if they didn't error. This way errors are caught early on.

Reply:
Sounds good, thanks for the feedback. I just pushed the validation. *Note* that the host inside a namespace could be something like `abc` that k8s will resolve, so I only check that the host `is not empty` string.